incubator/hnc/api/v1alpha2/hnc_config_conversion.go

```diff
@@ -0,0 +1,190 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	"sigs.k8s.io/multi-tenancy/incubator/hnc/api/v1alpha1"
+)
+
+// ConvertTo converts this hncconfig to the Hub version (v1alpha1).
+func (src *HNCConfiguration) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*v1alpha1.HNCConfiguration)
+
+	// Spec
+	sspts := src.Spec.Types
+	dspts := []v1alpha1.TypeSynchronizationSpec{}
+	for _, sspt := range sspts {
+		dspt := v1alpha1.TypeSynchronizationSpec{}
+		dspt.APIVersion = sspt.APIVersion
+		dspt.Kind = sspt.Kind
+		switch sspt.Mode {
+		case Propagate:
+			dspt.Mode = v1alpha1.Propagate
+		case Ignore:
+			dspt.Mode = v1alpha1.Ignore
+		case Remove:
+			dspt.Mode = v1alpha1.Remove
+		// The following are the invalid modes in v1alpha2, so they should remain
+		// invalid in v1alpha1. See https://github.com/kubernetes-sigs/multi-tenancy/issues/885
+		case SynchronizationMode(v1alpha1.Propagate):
+			dspt.Mode = v1alpha1.SynchronizationMode(Propagate)
+		case SynchronizationMode(v1alpha1.Ignore):
+			dspt.Mode = v1alpha1.SynchronizationMode(Ignore)
+		case SynchronizationMode(v1alpha1.Remove):
+			dspt.Mode = v1alpha1.SynchronizationMode(Remove)
+		default:
+			dspt.Mode = v1alpha1.SynchronizationMode(sspt.Mode)
+		}
+		dspts = append(dspts, dspt)
+	}
+	dst.Spec.Types = dspts
+
+	// Status.Types
+	sstts := src.Status.Types
+	dstts := []v1alpha1.TypeSynchronizationStatus{}
+	for _, sstt := range sstts {
+		dstt := v1alpha1.TypeSynchronizationStatus{}
+		dstt.APIVersion = sstt.APIVersion
+		dstt.Kind = sstt.Kind
+		dstt.NumSourceObjects = sstt.NumSourceObjects
+		dstt.NumPropagatedObjects = sstt.NumPropagatedObjects
+		switch sstt.Mode {
+		case Propagate:
+			dstt.Mode = v1alpha1.Propagate
+		case Ignore:
+			dstt.Mode = v1alpha1.Ignore
+		case Remove:
+			dstt.Mode = v1alpha1.Remove
+		default:
+			dstt.Mode = v1alpha1.SynchronizationMode(sstt.Mode)
+		}
+		dstts = append(dstts, dstt)
+	}
+	dst.Status.Types = dstts
+
+	// Status.Conditions
+	sstcs := src.Status.Conditions
+	dstcs := []v1alpha1.HNCConfigurationCondition{}
+	for _, sstc := range sstcs {
+		dstc := v1alpha1.HNCConfigurationCondition{}
+		dstc.Code = v1alpha1.HNCConfigurationCode(sstc.Code)
+		dstc.Msg = sstc.Msg
+		dstcs = append(dstcs, dstc)
+	}
+	dst.Status.Conditions = dstcs
+
+	// Status.NamespaceConditions
```

Comment:
Are we keeping NamespaceConditions? I have a funny memory that we weren't?

Reply:
We didn't have a conclusion on whether to keep or delete namespace conditions in HNCConfig, but only that if we have this condition, make sure we keep it rate-limited and length-limited. But someone did mention that users are more likely to find namespace conditions from namespaces (not somewhere else like in this config resource) to fix them.

So it's more like our call if we want to keep it or not. The purpose is to have all the conditions in one place for user to look. It's generated from forest, right? Do you think we can remove it since we already have metrics showing the same data now? (The metrics are generated from forest not from here, right?)

Reply:
I forget where the metrics come from but they don't identify _which_ namespaces have the problem. Hmm. If an admin has 1,000 namespaces, one of which has a condition, how would they find it if we don't have this?

Reply:
But you can ignore that question for the purpose of this PR :)

Reply:
sg, I will have it in its own PR if we decide a change there.